### PR TITLE
Redirect media directly to a Cloudinary URL

### PR DESCRIPTION
**Pull request by colbyfayock**
This updates plugin so that instead of using a serverless function as a middle step between redirects, it creates redirects directly to a Cloudinary URL.

When using the fetch delivery type, we're able to do this with a wildcard match as we're really just passing through the remote URL.

In the event we're using the upload delivery type, we're currently adding a redirect for each asset (inspired by the solution from Essential Next.js Plugin) as a way to still provide the redirect capability without further degrading performance with the serverless function in between. Though not ideal having that many additional redirects, it helps speed things a lot reducing the hops.

### CODE REVIEW

**Comment by ascorbic** [comment]
Rather than adding rewrites, could you replace the URL in the HTML? That would save a request, and would prevent the need for so many redirects in the config. It would also avoid potential issues with redirect precedence.

**Comment by colbyfayock** [comment]
> Rather than adding rewrites, could you replace the URL in the HTML? That would save a request, and would prevent the need for so many redirects in the config. It would also avoid potential issues with redirect precedence.

I’m actually already doing that on post build 😅 but clientside routing and images loaded via js aren't covered by that alone

**Comment by ascorbic** [comment]
Aah, ok missed that!

**Comment by ascorbic** [comment]
We do it because we have no choice, and would very much rather not do it.

**Comment by ascorbic** [comment]
If there was a way to create a wildcard redirect that could do it. e.g.

`/images/* https://res.cloudinary.com/demo/image/upload/:splat 302`

**Comment by colbyfayock** [comment]
oof yeah great point, that worked, just pushed that out

thinking out loud... that might only work reliably for using the Cloudinary `fetch` API though 🤔 

the upload API, the idea would be i would upload that asset to cloudinary programmatically during that build process then deliver it. the issue there, is i would then have a Public ID (cloudinary's ID) that i would need to use to deliver the image

theoretically, i could use the image URL as the public ID and the redirect would be basically the wildcard iteration, but im wondering if there would be any validation issues we'd run into 🤔 

say someone runs it, then updates an image using the same name, idk if it would properly serve that image? ill test

**Comment by colbyfayock** [comment]
Set it up so that if the delivery type is fetch, we can provide the wildcard-based matching, otherwise if it's upload, for now, we're mapping every asset, as i can't think of a way to do that with wildcard matching without adding the function inbetween with each image having its own unique public ID

https://github.com/colbyfayock/netlify-plugin-cloudinary/pull/12/files#diff-bfe9874d239014961b1ae4e89875a6155667db834a410aaaa2ebe3cf89820556R123

File: src/index.js [hunk 1]
```diff
@@ -20,109 +19,186 @@ const CLOUDINARY_MEDIA_FUNCTIONS = [
...
-    const { FUNCTIONS_SRC, INTERNAL_FUNCTIONS_SRC } = constants;
     const {
       deliveryType,
-      folder = process.env.SITE_NAME,
       uploadPreset,
+      folder = process.env.SITE_NAME
     } = inputs;
 
+    // If we're using the fetch API, we don't need to worry about uploading any
+    // of the media as itw ill all be publicly accessible, so we can skip this step
```
Comment (Chuloo) [inline src/index.js:40]: Typo in comment :)

File: src/index.js [hunk 2]
```diff
@@ -139,6 +215,7 @@ module.exports = {
       const sourceHtml = await fs.readFile(page, 'utf-8');
 
       const { html, errors } = await updateHtmlImagesToCloudinary(sourceHtml, {
+        assets: netlifyConfig.build.environment.CLOUDINARY_ASSETS,
```
Comment (Chuloo) [inline src/index.js:218]: Nice abstraction 👍

**Review by Chuloo — APPROVED** [review]
Looks neat!

**Pull request merged**